linkerd/stack/src/proxy.rs

```diff
@@ -39,3 +51,45 @@ where
         inner.call(req)
     }
 }
+
+#[pin_project::pin_project]
+pub struct Oneshot<P, S, R>
+where
+    P: Proxy<R, S>,
+    S: tower::Service<P::Request>,
+{
+    req: Option<R>,
+    #[pin]
+    state: State<P, S, P::Future>,
+}
+
+#[pin_project::pin_project(project = StateProj)]
+enum State<P, S, F> {
+    PollReady { proxy: P, svc: S },
+    Future(#[pin] F),
+}
+
+impl<P, S, R> Future for Oneshot<P, S, R>
+where
+    P: Proxy<R, S>,
+    S: tower::Service<P::Request>,
+    P::Error: From<S::Error>,
+{
+    type Output = Result<P::Response, P::Error>;
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let mut this = self.project();
+        loop {
+            match this.state.as_mut().project() {
+                StateProj::PollReady { proxy, svc } => {
+                    futures::ready!(svc.poll_ready(cx))?;
+                    let f = proxy.proxy(svc, this.req.take().expect("already called"));
+                    this.state.as_mut().set(State::Future(f));
+                }
+                StateProj::Future(f) => return f.poll(cx),
+            }
+        }
+    }
+}
```

Comment:
happy to factor this change out into a smaller branch that can merge independently